### ISSUE TITLE
Start loading `wait"1"` immediately after all with `wait"0"` loaded?

### ISSUE THREAD

**Issue by lainiwa** [zdharma/zplugin#134]
**Is your feature request related to a problem? Please describe.**
I do not see much use case for any numeric values of `wait` modifier other than `wait"0"`.
For now zplugin with different `wait` values would:
* first set the prompt (and other immediate stuff to set up)
* synchronously load plugins with `wait"0"` ice modifier in the order specified in config
* wait up to 1 second
* load plugins (if any) with `wait"1"` modifier
* ... and so on

As I see it, I think `wait` is expected to be used to specify the **order** rather then the exact number of **seconds** to wait before loading.

**Describe the solution you'd like**
As `wait"0"` means "immediately after prompt load", it might be more convenient to define `wait"1"` as "immediately after all plugins with `wait"0"` loaded".
This would let the user to have his configuration fully loaded as fast as possible while preserving the desired order.

**Comment by psprint** [comment]
That's an interesting idea. Currently, there is an automatic division of a single time-slot wait'X' – maximum of 20 events will get executed, the remaining ones will be deferred to the next time-slot.

But the idea means something different. It means to divide single time-slot into multiple event-firing sessions. This way, the prompt would not block even for large number of events in them in total.

That's an interesting, actually a ground-breaking change that would help to deliver the promise of loading in background, without interference. I'll have to think about it for a little (it seems to be completely doable at first sight) and find time to implement it.

**Comment by psprint** [comment]
Almost an miracle resolution – the automatic division now re-schedules the events to be run **immediately after the current firing-session.** In other words:
 * a fact: the prompt is being blocked inside a single firing session,
 * by re-scheduling the events into another firing-session we volunteer CPU time to the prompt – the prompt is being active in between the sessions
 * the new firing-session is being run instantly after the current one, so there is no gap and the events are being fired continuously

I've lowered the automatic-dividing number of events to 5. So in another words: **the prompt is being activated after each 5 events automatically**.

**Comment by lainiwa** [comment]
I updated zplugin, but I don't really notice the difference, heh.

Let me illustrate what I envisioned. Consider the following made-up code:
```sh
mkdir -p "${HOME}/.zplugin/plugins/_local---null" >/dev/null
zplugin ice lucid wait"0" atload"echo Wave 0 plugin 00"; zload _local/null
zplugin ice lucid wait"1" atload"echo Wave 1 plugin 10"; zload _local/null
zplugin ice lucid wait"1" atload"echo Wave 1 plugin 11"; zload _local/null
zplugin ice lucid wait"2" atload"echo Wave 2 plugin 20"; zload _local/null
zplugin ice lucid wait"2" atload"echo Wave 2 plugin 21"; zload _local/null
zplugin ice lucid wait"3" atload"echo Wave 3 plugin 30"; zload _local/null
zplugin ice lucid wait"3" atload"echo Wave 3 plugin 31"; zload _local/null
```

The code is loaded in about 3 seconds, while it probably **could** be load almost instantly (if we see 'wait' modifier as a purely setting the **order** rather than **time in seconds**).

In the code ^ the user specifies, that he wants plugins 1X be loaded strictly before plugins 2X, for example.
While the order of inside the "wave" (the time slot, as you call it) is probably not that important and might change when the user rearranges the lines.

Here is the reason why I think the **order of loading** should not be set by the **order of the lines**: the user will probably choose to have commands and their completions be located close to one another in his zshrc, while still being loaded in different waves (time slots), e.g.:
```sh
zplugin ice wait"1" ... ; zplugin load <some_commands>
zplugin ice wait"0" ... ; zplugin snippet <the_commands_completions>
```

As I understand the changes you made - is that now zplugin re-draws the prompt more frequently (every 5 (why not every 1, by the way?) loaded plugins instead of every 20) in between loading the plugins. Which is nice, but not the exact thing I meant. However, I'm confused by this statement
> the new firing-session is being run instantly after the current one, so there is no gap and the events are still being fired continuously

as the code I listed loades in 3 seconds, while the statement (as I understand it) implies it should be loaded instatnly. Am I getting something wrong?

**Comment by psprint** [comment]
A fast reply: use `wait'0`' or in general `wait'X'` (i.e. the same `wait''` argument) – the packs of fives are gathered among the same `wait''`-arguments.

**Comment by psprint** [comment]
As I read more thoroughly through your post, I have an opinion that the change from treating the `wait'X'` argument from **(A) time slot** to **(B) order of loading number** is too complex to the end user. I already was thinking about adding **(C) order of loading letter** to the (A) time slot, i.e. e.g.: `wait'1a'`, `wait'1b'`, etc. and I might still go on with this, however, is it possible to see this feature in a real-world example? I.e. a longer block of `zplugin` commands? To see the benefits that you mention, i.e.:

> the user will probably choose to have commands and their completions be located close to one another in his zshrc, while still being loaded in different waves (time slots), e.g.:

As for the current feature, to see that it is active and working, edit `-zplg-scheduler` function in a live Zsh-session by using `zed`:

```zsh
autoload zed; zed -f -zplg-scheduler
```

Add `echo rescheduled` after the `sched +0 -zplg-scheduler following-additional`  line:

```zsh
...
[[ $(( ++__idx, __count += ${${REPLY:+1}:-0} )) -gt 5 && "$1" != "burst" ]] && {
    sched +0 -zplg-scheduler following-additional
    echo 'rescheduled'
    break
}
...
```

and then execute following code:

```zsh
for (( i=0; i<=22; i ++ )); do
    zplg ice as"command" pick"wd.sh" atinit"echo Firing $i" wait'1' lucid
    zplg light mfaerevaag/wd
done
```

As for why I've chosen the number `5` as the size of the single firing-session pack, it might be irrational, indeed. I've just didn't want to stress `sched` that much, i.e. to reschedule that often. But provided that `sched` has no bugs, it should work without problem also with the pack size set to `1`. Let's see if users report any problems with the size of `5` and then change it to `1` :tada:

**Comment by psprint** [comment]
Ok it wasn't that difficult to code, so here it is the `wait'0b'`, `wait'0a'` support.

One can issue 2 zplugin commands, like:

```zsh
zplugin ice as"command" pick"wd.sh" atinit"echo Fireing 1" wait"0b" lucid
zplugin light mfaerevaag/wd
zplugin ice as"command" pick"wd.sh" atinit"echo Fireing 2" wait"0a" lucid
zplugin light mfaerevaag/wd
```

and have the second one ran before the first one, all in slot 0.

**Comment by lainiwa** [comment]
Sorry for being unresponsive for a while.

> the change from treating the wait'X' argument from (A) time slot to (B) order of loading number is too complex to the end user

I think there is some misunderstanding having place, as the way I see it, changing **(A)** to **(B)** would only make the loading continuous by reducing the gaps (when all events in time-slot N'th has finished loading, the is no reason not to start executing events in N+1'th slot).

While this change would break no current users' configurations, it would also allow users to re-implement the **(C)** scheme by setting `wait'11'` and `wait'12'` instead of `wait'1a'` and `wait'1b'` - as the fact that gaps between time-slots (or waves) would be removed (or it can be seen as time-slots being shrinked to their minimal necessary duration), it would be fine setting big numbers like 11 or 12 for wait modifier.

> I.e. a longer block of zplugin commands? To see the benefits that you mention

Consider my own [config](https://github.com/lainiwa/dotfiles/blob/ca80ad649c07a62dedbd6705d0931bb2cf1bfdf5/zsh/zplugin.zsh#L61-L72). I am placing events for loading commands and sourcing their completions nearby (as I guess most zplugin's users would do). However, I choose to load commands after their completions (as I need completions while I'm typing, but the commands would be executed only after I press Enter). If i would choose to go with more sophisticated setup and would use higher numbers in `wait` modifier, I might end up with a command-not-found error, for example.

I've played around modifying `-zplg-scheduler` a little, and contrarily to my initial belief it did not re-draw the prompt upon rescheduling. Thus it seems to be a very internal thing, not sure how it affects the end user. Setting different pack sizes does not change the loading behavior in any noticeable (without echo'ing "rescheduled") way either.

I feel a little weird that i fail to explain myself so far. I might be missing some technical detail that seems obvious to you, heh. I tried to illustrate my idea with a scheme of loading process:

![untitled](https://user-images.githubusercontent.com/26632061/58712270-f21a0480-83c8-11e9-875f-6253d0756e0a.png)

It is related to the code example I wrote above
> `mkdir -p "${HOME}/.zplugin/plugins/_local---null" >/dev/null`
> `zplugin ice lucid wait"0" atload"echo Wave 0 plugin 00"; zplugin load _local/null`
> `zplugin ice lucid wait"1" atload"echo Wave 1 plugin 10"; zplugin load _local/null`
> `zplugin ice lucid wait"1" atload"echo Wave 1 plugin 11"; zplugin load _local/null`
> `zplugin ice lucid wait"2" atload"echo Wave 2 plugin 20"; zplugin load _local/null`
> `zplugin ice lucid wait"2" atload"echo Wave 2 plugin 21"; zplugin load _local/null`
> `zplugin ice lucid wait"3" atload"echo Wave 3 plugin 30"; zplugin load _local/null`
> `zplugin ice lucid wait"3" atload"echo Wave 3 plugin 31"; zplugin load _local/null`

So my idea is basically to get rid of the red parts of time line.

**Comment by psprint** [comment]
Ok I understand what you mean. I've had the picture before. However I was thinking: "if the user wants to effectively divide the time slots even to the extent of abolishing the time slots completely in favor of single, maximum-fast time slot, then why 3-4 sub-slots per single time-slot would not be satisfactory? I know it's a big step-back from the single-wave, but isn't it just satisfactory?".

I'm getting a little inspired by the idea of single max-speed wave, I might code something and provide it as an option.

**Comment by psprint** [comment]
It seems that to implement the single-wave solution, it is sufficient to add the `+1000` below in the `-zplg-scheduler`:

```zsh
    integer __t=EPOCHSECONDS+1000 __i correct=0
```

I've started multiple sessions and the results seem good :)

**Comment by lainiwa** [comment]
Oh, using multiple `wait` modifiers with the same order-setting letter just did not come to my mind.
I've tried both ways (setting it with `wait'0a-z'` and modifying `-zplg-scheduler`, and both seem to work fine.
Cool! I think this should be very useful.

Closing the issue now.

**Comment by psprint** [comment]
BTW. thanks for sharing the `zshrc`, it has many interesting solutions.

**Comment by psprint** [comment]
@lainiwa: Could you consider adding your `zplugin.zsh` into the new [zplugin-configs](https://github.com/zdharma/zplugin-configs) repository? I can add it myself if you don't wan't to bother with a PR if you give me a permission.

**Comment by HaleTom** [comment]
Is this feature documented anywhere?  Could the main page `wait` doco at least refer to this issue?

**Comment by psprint** [comment]
@HaleTom You're referring to `a`,`b`,`c` suffixes?
